src/Java.Interop.Tools.JavaTypeSystem/Adapters/ManagedApiImporter.cs

```diff
@@ -0,0 +1,379 @@
+					var jni = (string) reg_attr.ConstructorArguments [0].Value;
+					var name = jni.Replace ('/', '.').Replace ('$', '.');
+
+					model.Implements.Add (new JavaImplementsModel (name, name, $"L{jni};"));
+				}
+			}
+		}
+
+		static string GetBaseTypeJni (TypeDefinition type)
+		{
+			// Find a Java base type, ignoring generic types, if nothing else it will be Java.Lang.Object
+			TypeDefinition? base_type = type;
+
+			while (true) {
+				base_type = base_type.BaseType?.Resolve ();
+
+				if (base_type is null)
+					break;
+
+				// These are the base types for Java.Lang.Object and Java.Lang.Throwable
+				if (base_type.FullName == "System.Object" || base_type.FullName == "System.Exception")
+					return string.Empty;
+
+				if (base_type.HasGenericParameters || base_type.IsGenericInstance)
+					continue;
+
+				if (GetRegisterAttribute (base_type.CustomAttributes) is CustomAttribute reg_attr)
+					return (string) reg_attr.ConstructorArguments [0].Value;
+			}
+
+			return "java/lang/Object";
+		}
+
+		static CustomAttribute? GetObsoleteAttribute (Collection<CustomAttribute> attributes) =>
+			attributes.FirstOrDefault (a => a.AttributeType.FullNameCorrected () == "System.ObsoleteAttribute");
+
+		static CustomAttribute? GetRegisterAttribute (Collection<CustomAttribute> attributes) =>
+			attributes.FirstOrDefault (a => a.AttributeType.FullNameCorrected () == "Android.Runtime.RegisterAttribute");
+
+		static string? GetRegisteredJavaTypeName (TypeDefinition type)
+		{
+			if (GetSpecialCase (type) is string s)
+				return s;
+
+			if (GetRegisterAttribute (type.CustomAttributes) is CustomAttribute reg_attr)
+				return ((string) reg_attr.ConstructorArguments [0].Value).Replace ('/', '.');
+
+			return null;
+		}
+
+		static string? GetSpecialCase (TypeDefinition type)
+		{
+			return type.FullName switch {
+				"System.Collections.Generic.IList`1" => "java.util.List",
+				"System.Collections.Generic.IDictionary`2" => "java.util.Map",
+				"System.Collections.Generic.ICollection`1" => "java.util.Collection",
+				"System.String" => "java.lang.String",
+				_ => null,
+			};
+		}
+
+		static string FullNameCorrected (this TypeReference t) => t.FullName.Replace ('/', '.');
+
+		public static string Visibility (this MethodDefinition m) =>
+	m.IsPublic ? "public" : m.IsFamilyOrAssembly ? "protected internal" : m.IsFamily ? "protected" : m.IsAssembly ? "internal" : "private";
+
+		static (string package, string nestedName) DecodeRegisterJavaFullName (string value)
+		{
+			var idx = value.LastIndexOf ('.');
+
+			var package = idx >= 0 ? value.Substring (0, idx) : string.Empty;
+			var nested_name = idx >= 0 ? value.Substring (idx + 1) : value;
+
+			return (package, nested_name);
+		}
+
+		static string FormatJniSignature (string package, string nestedName)
+		{
+			if (package.HasValue ())
+				return $"L{package.Replace ('.', '/')}/{nestedName.Replace ('$', '/')};";
```

Comment:
This doesn't work properly for types in the global package, e.g. `[Register ("Example")]`.

Additionally, `FormatJniSignature()`, as called from e.g. `ParseClass()`, is given outputs from `DecodeRegisterJavaFullName()`, which -- currently? -- is the first `[Register]`  constructor argument, e.g. `android/app/AlarmManager$OnAlarmListener`.

We *shouldn't* replace `$` with `/`, as that would result in a JNI name which doesn't exist: `android/app/AlarmManager/OnAlarmListener`.